sdk/helper/storagepacker/invalidate.go

```diff
@@ -0,0 +1,295 @@
+//
+// For each bucket B we receive, process the items in it, compared with items in
+// the previous version of B in the cache (if any, call it C) and elsewhere in the current cache.
+//
+// If an item exists in B, report it in "present" even if it also exists elsewere in the cache
+//
+// If an item exists in C, but not in B, then it might have been deleted.
+//    * If the item exists elsewhere in the cache (might be a child, might even be a parent)
+//      don't mark it as deleted. Either we will do it later (when we process that bucket)
+//      or the item is now in its correct location.
+//         * One problem here is that we can't rely upon the longest suffix match to find the bucket.
+//         * Another problem is that the local side might have a newer (or older) version in its
+//           memory compared to storage. If we can't be confident that InvalidateItems() will be
+//           later called on the bucket that *does* have the consistent version, we have to report any
+//           such item as "present" to give the upper layer a chance to look at it.
+//    * If the item does not exist elsewhere in the cache, report it as deleted.
+//
+// If an item has been moved by sharding, it's possible this will cause it to be reported
+// as deleted and re-created.
+//
+// Either B or C might be nonexistent; just treat those as empty.
+//
+// If we're the secondary, we shouldn't have to worry about concurrent access.
+func (s *StoragePackerV2) InvalidateItems(ctx context.Context, path string, newValue []byte) (present []*Item, deleted []*Item, err error) {
+	// Double-check that it's on our path, and figure out which bucket it should be
+	bucketKey, err := s.bucketKeyFromPath(path)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var replacementBucket *LockedBucket
+	bucketRemoved := false
+
+	if newValue == nil {
+		bucketRemoved = true
+		// Create a placeholder so we can run the rest of the logic
+		// against an empty collection
+		replacementBucket = s.newEmptyBucket(bucketKey)
+	} else {
+		// Fake up a storage entry for the bucket and decode it
+		storage := &logical.StorageEntry{
+			Value: newValue,
+		}
+		replacementBucket, err = s.DecodeBucket(storage)
+		if err != nil {
+			return nil, nil, errwrap.Wrapf("unable to decode replicated bucket: {{err}}", err)
+		}
+		if replacementBucket.Key != bucketKey {
+			return nil, nil, fmt.Errorf("decoded bucket key %q doesn't match path component %q", replacementBucket.Key, bucketKey)
+		}
+	}
+
+	present = make([]*Item, 0)
+
+	originalBucket, err := s.GetBucket(ctx, bucketKey)
+	if err != nil {
+		return nil, nil, errwrap.Wrapf("problem finding original bucket: {{err}}", err)
+	}
+	originalBucket.Lock()
+	defer originalBucket.Unlock()
+
+	// Any item included in the new bucket should be reported present.
+	if !replacementBucket.HasShards && replacementBucket.ItemMap != nil {
+		for k, v := range replacementBucket.ItemMap {
+			present = append(present, &Item{
```

Comment:
Because of how `present` is instantiated, I think it needs to be inserted at an index like [this](https://play.golang.org/p/CVyezDGOlX1). However, if you don't want to track the index, and you don't need to optimize memory here, you could instead just instantiate `present` like this: 
```
var present []string
```
That'll grow it in size automatically as you append things, though I _think_ it has poorer memory performance.

Reply:
FWIW, though I haven't tested lately, I _think_ that if you use a slice in a loop like this:
```
var slc []string
for {
    slc = append(slc, thing)
}
```
Then on each iteration of the loop it actually creates a new pointer to a new slice with a different (incremented) size. When this happens in a loop, I _think_ Go's garbage collector won't collect the slices until the method is over so `slc` isn't in scope anymore. 

I could totally be wrong but I could swear I saw something like that in earlier versions of Go. If memory is important here, might be worth a double-check.

Reply:
I'm not clear on what the problem is. When I run your snippet with size 0 (like I used here) it seems fine.  I couldn't decide whether it was worth pre-allocating for the worst case (either all objects present, or all objects deleted.)

In this particular case, present is already declared (as a return parameter) but if 
```
var deleted []*Item
```
is more idiomatic or better-performing then I'm happy to switch.

Reply:
You're totally right. For some reason I thought the fixed size wasn't 0.